lightclient/src/platform/wasm_socket.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use futures::{io, prelude::*};
+use send_wrapper::SendWrapper;
+use wasm_bindgen::{prelude::*, JsCast};
+
+use std::{
+    cell::RefCell,
+    collections::VecDeque,
+    pin::Pin,
+    rc::Rc,
+    task::Poll,
+    task::{Context, Waker},
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Failed to connect {0}")]
+    ConnectionError(String),
+}
+
+/// Websocket for WASM environments.
+///
+/// This is a rust-based wrapper around browser's WebSocket API.
+pub struct WasmSocket {
+    /// Inner data shared between `poll` and web_sys callbacks.
+    ///
+    /// This is safe in wasm environments.
+    inner: SendWrapper<Rc<RefCell<InnerWasmSocket>>>,
```

Comment:
Ah, so the reason for the `Arc<Mutex<..>>` was because this thing needed sharing across some different closures below; I missed that bit! And when compiled to WASM the mutex will disappear anyway since it's single threaded.

I think given that the outer struct is not cloneable, I'd remove the `SendWrapper` thing (and crate) and just put the `Arc<Mutex<..>>` back (it feels cleaner than this `SendWrapper<Rc<RefCell<..>>>` thing). Just maybe a dev comment on `WasmSocket` that the thing cannot impl `Clone` would do I think :)